core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceRedundantJoinWithSource.java

```diff
@@ -35,18 +53,61 @@
     @Override
     public Result apply(JoinNode node, Captures captures, Context context)
     {
-        if (node.getRight().getOutputSymbols().isEmpty() && isScalar(node.getRight(), context.getLookup())) {
-            return Result.ofPlanNode(node.getLeft());
+        if (isAtMost(node.getLeft(), context.getLookup(), 0) || isAtMost(node.getRight(), context.getLookup(), 0)) {
+            return Result.empty();
         }
-        if (node.getLeft().getOutputSymbols().isEmpty() && isScalar(node.getLeft(), context.getLookup())) {
-            return Result.ofPlanNode(node.getRight());
+
+        boolean leftSourceScalarWithNoOutputs = node.getLeft().getOutputSymbols().isEmpty() && isScalar(node.getLeft(), context.getLookup());
+        boolean rightSourceScalarWithNoOutputs = node.getRight().getOutputSymbols().isEmpty() && isScalar(node.getRight(), context.getLookup());
+
+        switch (node.getType()) {
+            case INNER:
+                if (node.isCrossJoin()) {
+                    if (leftSourceScalarWithNoOutputs) {
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), node.getRight(), ImmutableSet.copyOf(node.getRightOutputSymbols()))
+                                .orElse(node.getRight()));
+                    }
+                    if (rightSourceScalarWithNoOutputs) {
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), node.getLeft(), ImmutableSet.copyOf(node.getLeftOutputSymbols()))
+                                .orElse(node.getLeft()));
+                    }
+                }
+                else {
+                    if (leftSourceScalarWithNoOutputs) {
+                        FilterNode filterNode = new FilterNode(context.getIdAllocator().getNextId(), node.getRight(), node.getFilter().get());
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), filterNode, ImmutableSet.copyOf(node.getRightOutputSymbols()))
+                                .orElse(filterNode));
+                    }
+                    if (rightSourceScalarWithNoOutputs) {
+                        FilterNode filterNode = new FilterNode(context.getIdAllocator().getNextId(), node.getLeft(), node.getFilter().get());
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), filterNode, ImmutableSet.copyOf(node.getLeftOutputSymbols()))
+                                .orElse(filterNode));
+                    }
+                }
+                break;
```

Comment:
This can be simplified to 

```java
PlanNode source;
List<Symbol> sourceOutputs;
if (leftSourceScalarWithNoOutputs) {
    source = node.getRight();
    sourceOutputs = node.getRightOutputSymbols();
}
else if (rightSourceScalarWithNoOutputs) {
    source = node.getLeft();
    sourceOutputs = node.getLeftOutputSymbols();
}
else {
    return Result.empty();
}

if (node.getFilter().isPresent()) {
    source = new FilterNode(context.getIdAllocator().getNextId(), source, node.getFilter().get());
}

return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), source, ImmutableSet.copyOf(sourceOutputs)).orElse(source));
```